django/db/models/expressions.py

```diff
@@ -1252,6 +1255,28 @@ def __repr__(self):
         return "{}({})".format(self.__class__.__name__, self.expression)
 
 
+class NegatedExpression(ExpressionWrapper):
+    """
+    Wrapping an expression to negate its output.
+    """
+
+    def __init__(self, expression):
+        super().__init__(expression, output_field=fields.BooleanField())
+
+    def __invert__(self):
+        return copy.copy(self.expression)
+
+    def as_sql(self, compiler, connection):
+        try:
+            sql, params = super().as_sql(compiler, connection)
+        except EmptyResultSet:
+            features = compiler.connection.features
+            if not features.supports_boolean_expr_in_select_clause:
+                return "1=1", ()
+            return compiler.compile(Value(True))
+        return f"NOT {sql}", params
```

Comment:
This behaves quite unexpected with database functions on some databases (SQLite or MySQL). For example:
`~ExtractYear("start_datetime")` crashes on PostgreSQL but returns `False` on SQLite and MySQL :open_mouth: I'm not sure if we can and should do anything about it :thinking:

Reply:
We should probably forbid these cases, as @charettes suggested above.
I added `Negated.resolve_expression` which checks on the resolved expression if it's conditional.